UNL-CSE-AMBASSADORS/VITA
# Self-Service Client Reschedule and Cancel

This PR enables clients to reschedule or cancel their own appointments without having to contact the center. I could've broken this feature into a few PRs, but oh wells ¯\_(ツ)_/¯

I did a lot of code cleanup in this PR since this feature led to some duplication between the appointment management functionality and this self-service functionality. So you will see quite a bit of that in here as well.

**PLEASE NOTE**: This PR pulls in both #233 and #236, so it should **not** be merged until those two are merged. It also may be hard to review in isolation until #233 is merged in.

**TESTING**
1. Run the DB creation script and the test data scripts
2. Navigate to the signup page and signup for an appointment
    * NOTE: To do this you will need to uncomment the comments in signup/signup.php that are preventing one from signing up. This is lines 4, 49, 88, 264. Note those line numbers are before removing any of the comment markers
    * ALSO NOTE: Remember all the personal information you put in! It is required in a later step
3. In the appointment confirmation, there is a new header for rescheduling and it provides a link.
4. Navigate to that link from step 3
    * NOTE: We prepend `https://` in the code, but your localhost probably isn't set up to handle that, so if the page doesn't load, you will probably need to remove the `https://` part
5. Welcome to the client reschedule and cancel page. 
6. Enter in information that *DOES NOT* match what you had put in when signing up.
    * It should say the information didn't match our records and clear all the input fields
7. Enter in information that *DOES* match what you had put in when signing up.
    * It should now show the reschedule/cancel functionalities
8. Scroll down and reschedule the appointment to any time, note which time you change it to
9. It should succeed, give a notification saying it succeeded, and then show the familiar appointment confirmation screen. 
    * Scroll down and the reschedule heading and information should also be there (note that the link is expected to be the same)
    * The print button should open the print functionality of the browser
    * The email me button should successfully send an email (you can see the message in the network tab) and change to say "Sent!" and be disabled after you hit it
10. Verify in the database that your appointment was rescheduled (alternatively verify on the appt management page)
    * Also, go to your appointment on the appt mgmt page, and there should be an automatic note saying the appointment was rescheduled by the client
11. Navigate again to your client reschedule link.
12. Enter in the correct information for your appointment
13. The current appointment information should display the new re-scheduled time.
14. Hit the "Cancel" button
    * A modal should popup asking you to confirm your cancellation
        * Clicking outside the modal box should close it with no cancellation performed
        * Clicking the X at the top right should close it with no cancellation performed
        * Clicking the "No, dont cancel" button should close it with no cancellation performed
15. Click the "Yes, Cancel" button
    * It should succeed, display a green success notification, and say your appointment has been cancelled
16. Verify the ServicedAppointment associated with your appointment is marked as cancelled and complete = false (either through the DB or appt mgmt page)
    * On the appt mgmt page for your appointment, there should be an automatic note saying it was cancelled by the client
17. Navigate again to your client reschedule link
18. It should say that this appointment can't be reschedule/cancelled and give the reason that it has already been cancelled
19. In the database, delete the ServicedAppointment entry that had been created for you
20. On the private queue, "Check In" your appointment
21. Navigate again to your client reschedule link
    * It should say can't be rescheduled/cancelled because the appointment has already been started
22. In the database, set the `timeIn` column on your ServicedAppointment entry to `NULL`
23. In the database, set the `scheduledTime` column on the AppointmentTime entry associated with your appointment to a date in the past
24. Navigate again to your client reschedule link
    * It should say can't be rescheduled/cancelled because the appointment's scheduled time is in the past
25. Change the token to something invalid (cases below) and it should say you've reached this page in mistake and tell them to click the link to go back home: 
    * More than 32 characters
    * Less than 32 characters
    * A token that doesn't have an associated appointment (just change one character in yours)


That is generally all the testing for this feature. You could also testing the "failure" notifications by adding `die()` calls in some of those private methods in the server/appointment/clientReschedule. I will not detail instructions for this however.

**MISCELLANEOUS TESTING**
Once again, I touched a lot of things making this PR, so a few things should be verified that they still work.
* Appointment Reschedule from the appt mgmt page
    * You should still get a success notification when the reschedule is successful
* Appointment cancellation from the appt mgmt page
* Cancellation from the private queue

**NEW DATABASE ERD**
![databaseerd](https://user-images.githubusercontent.com/9295744/44295346-bc0df880-a26c-11e8-9e02-1ab291bb5b1c.png)
